pages/docs/reference/js-ir-compiler.md

```diff
@@ -0,0 +1,71 @@
+title: "Using the experimental IR compiler"
+---
+# Using the experimental Kotlin/JS IR compiler
+
+> The Kotlin/JS IR compiler is an experimental feature in Kotlin 1.4. All of the functionality, language
+and tooling features described in this document are subject to change in future Kotlin versions.
+{:.note}
+
+The (currently still experimental) Kotlin/JS IR compiler backend is the main focus of innovation around Kotlin/JS, and paves the way forward for the technology. 
+
+Rather than directly generating JavaScript code from Kotlin source code, the Kotlin/JS IR compiler backend leverages a new approach. Kotlin source code is first transformed into an intermediate representation (IR), which is subsequently compiled into JavaScript. This allows for more aggressive optimizations, and enables improvements on pain points that were present in the previous compiler. This includes the areas of developer experience, generated code size (dead code elimination), and JavaScript and TypeScript ecosystem interoperability, to name just a few.
+
+The experimental IR compiler backend is available starting with Kotlin 1.4 through the Kotlin/JS Gradle plugin. To enable it in your project, pass a compiler type to the `js` function in your Gradle build script:
+
+<!--suppress ALL -->
+<div class="sample" markdown="1" mode="groovy" theme="idea">
+
+```groovy
+kotlin {
+    js(IR) { // or: LEGACY, BOTH
+        // . . .
+    }
+    binaries.executable()
+}
+```
+
+</div>
+
+- `IR` uses the new IR compiler backend for Kotlin/JS.
+- `LEGACY` uses the default compiler backend.
+- `BOTH` compiles your project with the new IR compiler as well as the default compiler backend. This is mainly useful for authoring libraries that are compatible with both backends, see [below](#authoring-libraries-for-the-ir-compiler-with-backwards-compatibility).
+
+The compiler type can also be set in the `gradle.properties` file, with the key `kotlin.js.compiler=ir`. (This behaviour is overwritten by any settings in the `build.gradle(.kts)`, however).
+
+## Current limitations of the IR compiler
+
+A major change with the new IR compiler backend is the **absence of binary compatibility** with the default backend. A lack of such compatibility between the two backends for Kotlin/JS means that a library created with the new IR compiler backend can’t be used from the default backend, and vice versa.
+
+If you want to use the IR compiler backend for your project, you need to **update all Kotlin dependencies to versions that support this new backend**. Libraries published by JetBrains for Kotlin 1.4+ targeting Kotlin/JS already contain all artifacts required for usage with the new IR compiler backend.
+
+**If you are a library author** looking to provide compatibility with the current compiler backend as well as the new IR compiler backend, additionally check out the [“Authoring libraries for the IR compiler”](#authoring-libraries-for-the-ir-compiler-with-backwards-compatibility) section.
```

Comment:
I assume this refers to the following snippet in the "What's new section":

> You can use code compiled by the new JVM IR back-end only if you've enabled the new back-end. Otherwise, you will get an error. Considering this, we don't recommend that library authors switch to the new back-end in production.

With Kotlin/JS, we have the `both` mode, which allows you to create artifacts that are compatible with IR and the legacy backend. As such, Library authors can use `both` mode to already switch on IR support without giving up the "production-ready" backend in their artifacts. This is also hopefully clarified by the corresponding part below:

> If you're a library maintainer who is looking to provide compatibility with the default back-end as well as the new IR compiler back-end, a setting for the compiler selection is available that allows you to create artifacts for both back-ends, allowing you to keep compatibility for your existing users while providing support for the next generation of Kotlin compiler.

Reply:
Let's discuss whether it makes sense to highlight this difference in Unified back-ends what's new ( JS IR back-end has `both` mode while JVM back-end doesn't and that's why can't be used in production).

As far as I understand, we don't recommend that you use only new JS IR back-end in production. I would make it clear - don't use only new back-end in production and use `both` mode (maybe your description is too long and not straight forward only for me).

Reply:
With the expanded note on _Alpha_ stability level, I think it becomes clear for people that the technology is not production-ready just yet.